### PR TITLE
[Bugfix][KV Connector] Cap MooncakeStore lookup at num_tokens - 1

### PR DESCRIPTION
**Pull request by n-dlms**
---
## Summary

Caps the MooncakeStore lookup result at `num_tokens - 1` inside `MooncakeStoreScheduler.get_num_new_matched_tokens`, so a lookup that reports the full prompt can no longer drive `num_new_tokens` to zero in `Scheduler.schedule()` and trip `assert num_new_tokens > 0` (`vllm/v1/core/sched/scheduler.py:923`) on the sync (`load_async=False`) path.

The cap is applied before both uses of the value, the returned `need_to_allocate` and the cached `load_specs.kvpool_cached_tokens`, so the round trip through the scheduler back into `update_state_after_alloc` (which asserts `num_external_tokens == kvpool_cached_tokens - vllm_cached_tokens`) stays consistent. No scheduler code is touched.

## Background

The last prompt token must always be recomputed locally to obtain logits. That rule is already enforced in three places: the local prefix cache lookup (`get_computed_blocks`), five in tree connectors (example, HF3FS, LMCache v1 adapter, MoRIIO READ, decode bench), and the MooncakeStore worker itself, which rederives a full hit below the request end (`worker.py:1854`). The MooncakeStore scheduler side component was the one place that passed the raw lookup count through uncapped.

One precision note for reviewers: through a matched pair of in tree server and scheduler components, the worker side rederivation keeps the lookup below N today. This change caps the scheduler side component itself, so the assert gated sync path no longer depends on server side behavior. The scheduler side receives its number over ZMQ from a server the base connector contract does not bound, and any server returning N (a third party implementation, a custom `LookupKeyServer`, or a future refactor) would crash `schedule()` without this guard.

## The fix

`vllm/distributed/kv_transfer/kv_connector/v1/mooncake/store/scheduler.py` (+10 lines): `min(num_external_hit_tokens, max(request.num_tokens - 1, 0))` immediately after the lookup resolves and before `need_to_allocate` and `load_specs` are derived. The cap is idempotent against values the worker already truncated, preserves the async flag, and leaves every below cap result byte for byte unchanged.

## Why this is not duplicating an existing PR

| Search | Result |
|---|---|
| Open PRs: "mooncake store cap", "MooncakeStore load_async=False" | Only #51358 (Mamba boundary state saves, unrelated) |
| Open PRs: "connector num_tokens - 1 cap" | None touching this method |
| #50864 (mine, scheduler side clamp for the same assert) | Closed in favor of this PR, per the review there that the fix belongs in the connector |

## Regression tests

Both tests fail with the fix stashed and pass with it applied.

| Test | What it proves |
|---|---|
| `test_full_prompt_lookup_capped_at_num_tokens_minus_one` | Connector side, existing `_StubLookupClient` pattern: lookup 32, computed 0, `num_tokens` 32 returns 31, async flag preserved, `load_specs` caches 31 |
| `test_full_prompt_lookup_does_not_crash_scheduler` | End to end through the real `Scheduler` and the real `MooncakeStoreConnector` scheduler side with only the ZMQ boundary stubbed and `load_async=False`: 32 token prompt schedules exactly 1 local token, `num_computed_tokens` reaches 32 (31 external + 1 local) |

## Test results (CPU only box, AMD 2 core)

```
.venv/bin/python -m pytest tests/v1/kv_connector/unit/test_mooncake_store_scheduler.py -q -p no:cacheprovider
# 24 passed (22 pre-existing + 2 new)

.venv/bin/python -m pytest tests/v1/kv_connector/unit/test_mooncake_store_scheduler.py \
  tests/v1/kv_connector/unit/test_mooncake_store_connector.py \
  tests/v1/kv_connector/unit/test_mooncake_store_coordinator.py \
  tests/v1/kv_connector/unit/test_mooncake_store_prepare_values.py -q -p no:cacheprovider
# 91 passed

.venv/bin/python -m pytest tests/v1/kv_connector/unit/test_mooncake*.py -q -p no:cacheprovider
# 247 passed, 18 failed. The 18 are byte identical with and without this change (verified by stash flip on test names): pre-existing environment gaps (missing pytest-asyncio plugin, RDMA dependent worker paths).

.venv/bin/python -m ruff check <both touched files>   # All checks passed
.venv/bin/python -m ruff format --check <both files>  # already formatted
```

Additionally verified by direct calls: overshoot capped (32/32 to 31), sub cap values unchanged, `num_tokens=1` yields no phantom load, already truncated inputs pass through untouched (idempotent), and the async default path (`load_async=True`) still routes to `WAITING_FOR_REMOTE_KVS` as designed with consistent cached accounting.

## Limitations and what CI should cover

No GPU on this box, so no serving or accuracy evals were possible. The change is CPU pure accounting (`min` on a token count) and only alters behavior where the scheduler previously asserted, but GPU validation of the disaggregated prefill path is requested. mypy was skipped (times out past 120s on this machine; no new typing surface was added).

## AI assistance disclosure

Per repo policy this PR is AI assisted. A human submitter reviewed every changed line, ran the tests above locally, and takes responsibility for the change end to end.

Co-authored-by: Cline

### CODE REVIEW

File: vllm/distributed/kv_transfer/kv_connector/v1/mooncake/store/scheduler.py [hunk 1]
```diff
@@ -139,6 +139,16 @@ def get_num_new_matched_tokens(
             return None, False
         num_external_hit_tokens = lookup_result.hit_length
 
+        # Cap at num_tokens - 1: the last prompt token must be recomputed
+        # locally for logits, matching the idiom applied by example, hf3fs,
+        # lmcache v1, moriio-READ and the local get_computed_blocks. Without
+        # this, a sync (`load_async=False`) lookup covering the full prompt
+        # drives num_new_tokens==0 in the scheduler and trips
+        # assert num_new_tokens > 0 (scheduler.py:923).
+        num_external_hit_tokens = min(
```
Comment (rishabhsinha17) [inline vllm/distributed/kv_transfer/kv_connector/v1/mooncake/store/scheduler.py:148]: The only server this can change anything for is one that returns `hit_length >= num_tokens`, and for that server the capped length is not on an `align_lookup_length` boundary, while the `LoadSpec` below still carries the `tail_key_boundaries` that server computed for its uncapped hit. The worker then loads with `token_len = kvpool_cached_tokens` and takes the `end == token_len` substitution in `_handle_request`, indexing `block_hashes` with the stale boundary. So in the one scenario this guard exists for, the loud scheduler assert becomes a quiet mis-keyed tail load. I would mirror the server's own rederivation instead, floor to the `align` already computed at the top of this method and drop the returned `tail_key_boundaries` whenever the cap fires (an aligned end takes no substitute key). That is also how example and hf3fs apply the idiom, both floor after the -1.

File: vllm/distributed/kv_transfer/kv_connector/v1/mooncake/store/scheduler.py [hunk 2]
```diff
@@ -139,6 +139,16 @@ def get_num_new_matched_tokens(
             return None, False
         num_external_hit_tokens = lookup_result.hit_length
 
+        # Cap at num_tokens - 1: the last prompt token must be recomputed
+        # locally for logits, matching the idiom applied by example, hf3fs,
+        # lmcache v1, moriio-READ and the local get_computed_blocks. Without
+        # this, a sync (`load_async=False`) lookup covering the full prompt
+        # drives num_new_tokens==0 in the scheduler and trips
+        # assert num_new_tokens > 0 (scheduler.py:923).
```
Comment (rishabhsinha17) [inline vllm/distributed/kv_transfer/kv_connector/v1/mooncake/store/scheduler.py:147]: nit: the assert has already moved (it is around 1007 on this branch and 1046 on today's main), maybe just say `Scheduler.schedule()` so the reference doesn't rot.

File: vllm/distributed/kv_transfer/kv_connector/v1/mooncake/store/scheduler.py [hunk 3]
```diff
@@ -139,6 +139,16 @@ def get_num_new_matched_tokens(
...
+        num_external_hit_tokens = min(
+            num_external_hit_tokens, max(request.num_tokens - 1, 0)
```
Comment (rishabhsinha17) [inline vllm/distributed/kv_transfer/kv_connector/v1/mooncake/store/scheduler.py:149]: nit: the `request.num_tokens < align` early return above already guarantees `num_tokens >= 1`, so the `max(..., 0)` can't fire.

File: tests/v1/kv_connector/unit/test_mooncake_store_scheduler.py [hunk 4]
```diff
@@ -1393,3 +1393,94 @@ def test_worker_metadata_aggregates_completions_across_ranks():
...
+        request_id="req-0",
+        num_tokens=32,
+        block_hashes=[b"h0", b"h1"],
+    )
+
+    need_to_allocate, load_async = scheduler.get_num_new_matched_tokens(
+        request, num_computed_tokens=0
+    )
+
+    assert need_to_allocate == 31, (
```
Comment (rishabhsinha17) [inline tests/v1/kv_connector/unit/test_mooncake_store_scheduler.py:1419]: With the defaults here (block 16, no partial hits) 31 is a length the in-tree pair can neither return (the server floors 31 to 16) nor load at key granularity, and the e2e test below stubs the lookup at the ZMQ boundary so nothing checks that 31 is actually loadable. If the cap floors to `align` as suggested this becomes 16 and the test pins a reachable value.

**Review by rishabhsinha17 — COMMENTED** [review]
Verified the failure mode and the accounting here. On the sync path an uncapped full-prompt hit drives `num_computed_tokens` to `request.num_tokens` and `schedule()` trips the `assert num_new_tokens > 0` after the budget min, so the crash is real and connector-side is the right place to fix it. The round trip is also consistent, `need_to_allocate` and `kvpool_cached_tokens` both derive from the same capped value and the scheduler passes `num_external_computed_tokens` through unchanged into `update_state_after_alloc`, so the equality assert there holds without the special-casing the LMCache adapter needs.

For the in-tree pair this is a no-op as described. The lookup server already rederives a full hit through `align_lookup_length(num_tokens - 1)` before replying (it lives in `LookupKeyServer.lookup` around worker.py:2202 these days, not 1854), so anything the client sees is already floored to a hash or block boundary below N. Which is also my one real concern with the shape of the cap, see inline.

On the cited idiom, moriio READ and decode_bench match as described. example and hf3fs both floor after the -1 rather than using the raw value, and the LMCache v1 adapter is a different mechanism, it only decrements the returned count when the hit equals `num_tokens`, keeps `lmcache_cached_tokens` uncapped, and skips the equality assert for that case. So the closest precedent for capping the hit itself is example/hf3fs, and both align, which is what the inline comment is about.
